File: contents/verlet_integration/code/julia/verlet.jl
function verlet(pos::Float64, acc::Float64, dt::Float64)
    prev_pos = pos
    time = 0.0

    while (pos > 0)
        time += dt
        temp_pos = pos
        pos = pos * 2 - prev_pos + acc * dt * dt
        prev_pos = temp_pos
    end

    return time
end

function stormer_verlet(pos::Float64, acc::Float64, dt::Float64)
    prev_pos = pos
    time = 0.0
    vel = 0.0

    while (pos > 0.0)
        time += dt
        temp_pos = pos
        pos = pos * 2 - prev_pos + acc * dt * dt
        prev_pos = temp_pos

        # Because acceleration is constant, velocity is straightforward
        vel += acc * dt
    end

    return time, vel
end

function velocity_verlet(pos::Float64, acc::Float64, dt::Float64)
    prev_pos = pos
    time = 0.0
    vel = 0.0

    while (pos > 0.0)
        time += dt
        pos += vel * dt + 0.5 * acc * dt * dt;
        vel += acc * dt;
    end

    return time, vel
end

function main()
    time = verlet(5.0, -10.0, 0.01);
    println("[#]\nTime for Verlet integration is:")
    println("$(time)")

    time, vel = stormer_verlet(5.0, -10.0, 0.01);
    println("[#]\nTime for Stormer Verlet integration is:")
    println("$(time)")
    println("[#]\nVelocity for Stormer Verlet integration is:")
    println("$(vel)")
    
    time, vel = velocity_verlet(5.0, -10.0, 0.01);
    println("[#]\nTime for velocity Verlet integration is:")
    println("$(time)")
    println("[#]\nVelocity for velocity Verlet integration is:")
    println("$(vel)")

end

main()

File: contents/verlet_integration/code/golang/verlet.go
package main

import "fmt"

func verlet(pos, acc, dt float64) (time float64) {
	prevPos := pos
	time = 0

	for pos > 0 {
		time += dt
		nextPos := pos*2 - prevPos + acc*dt*dt
		prevPos, pos = pos, nextPos
	}

	return
}

func stormerVerlet(pos, acc, dt float64) (time, vel float64) {
	prevPos := pos
	time, vel = 0, 0

	for pos > 0 {
		time += dt
		vel += acc * dt
		nextPos := pos*2 - prevPos + acc*dt*dt
		prevPos, pos = pos, nextPos
	}

	return
}

func velocityVerlet(pos, acc, dt float64) (time, vel float64) {
	time, vel = 0, 0

	for pos > 0 {
		time += dt
		pos += vel*dt + .5*acc*dt*dt
		vel += acc * dt
	}

	return
}

func main() {
	time := verlet(5., -10., .01)
	fmt.Println("[#]\nTime for Verlet integration is:")
	fmt.Println(time)

	time, vel := stormerVerlet(5., -10., .01)
	fmt.Println("[#]\nTime for Stormer Verlet integration is:")
	fmt.Println(time)
	fmt.Println("[#]\nVelocity for Stormer Verlet integration is:")
	fmt.Println(vel)
	
	time, vel = velocityVerlet(5., -10., .01)
	fmt.Println("[#]\nTime for velocity Verlet integration is:")
	fmt.Println(time)
	fmt.Println("[#]\nVelocity for velocity Verlet integration is:")
	fmt.Println(vel)
}
